### PR TITLE
Fix GCC13 build errors

### LINKED ISSUE #1639: [Bug]: build of 23.1.5 with gcc 13 fails with overloaded-virtual
### Which component impacted?

Build

### Is it regression? Good in old configuration?

None

### What happened?

Build media-driver with gcc-13-20230402.
builds fine with gcc-12.2

### What's the usage scenario when you are seeing the problem?

Others

### What impacted?

Build fails using gcc-13 20230402
```
[155/1453] Building CXX object media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp.o
FAILED: 
…
media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp.o -MF media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp.o.d -o media_driver/CMakeFiles/iHD_drv_video_COMMON.dir/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp.o -c /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-12.0-devel/build/media-driver-23.1.5/media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp
In file included from ../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_generic.h:31,
                 from ../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.h:30,
                 from ../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.cpp:26:
../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_generic.h: In instantiation of 'class MhwVdboxAvpInterfaceGeneric<mhw_vdbox_avp_g12_X>':
../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_g12_X.h:38:40:   required from here
../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_interface.h:779:24: error: 'virtual MOS_STATUS MhwVdboxAvpInterface::AddAvpBsdObjectCmd(PMOS_COMMAND_BUFFER, PMHW_BATCH_BUFFER, MhwVdboxAvpBsdParams*)' was hidden [-Werror=overloaded-virtual=]
  779 |     virtual MOS_STATUS AddAvpBsdObjectCmd(
      |                        ^~~~~~~~~~~~~~~~~~
../media_driver/agnostic/gen12/hw/vdbox/mhw_vdbox_avp_generic.h:99:16: note:   by 'MOS_STATUS MhwVdboxAvpInterfaceGeneric<TAvpCmds>::AddAvpBsdObjectCmd(PMOS_COMMAND_BUFFER, MhwVdboxAvpBsdParams*) [with TAvpCmds = mhw_vdbox_avp_g12_X; MOS_STATUS = _MOS_STATUS; PMOS_COMMAND_BUFFER = _MOS_COMMAND_BUFFER*]'
   99 |     MOS_STATUS AddAvpBsdObjectCmd(
      |                ^~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

notes:
- https://gcc.gnu.org/bugzilla/show_bug.cgi?id=20423
- gcc-13 includes -Woverloaded-virtual

### Debug Information

_No response_

### Do you want to contribute a patch to fix the issue?

None

### LINKED ISSUE #1640: [Bug]: build of 23.1.5 with gcc 13 fails with dangling-pointer
### Which component impacted?

Build

### Is it regression? Good in old configuration?

None

### What happened?

Build media-driver with gcc-13-20230402.
builds fine with gcc-12.2

### What's the usage scenario when you are seeing the problem?

Others

### What impacted?

Build fails using gcc-13 20230402
```
In member function 'void std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::swap(std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>&) [with _Key = std::__cxx11::basic_string<char>; _Val = std::pair<const std::__cxx11::basic_string<char>, CMRTKernelBase*>; _KeyOfValue = std::_Select1st<std::pair<const std::__cxx11::basic_string<char>, CMRTKernelBase*> >; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, CMRTKernelBase*> >]',
    inlined from 'void std::map<_Key, _Tp, _Compare, _Alloc>::swap(std::map<_Key, _Tp, _Compare, _Alloc>&) [with _Key = std::__cxx11::basic_string<char>; _Tp = CMRTKernelBase*; _Compare = std::less<std::__cxx11::basic_string<char> >; _Alloc = std::allocator<std::pair<const std::__cxx11::basic_string<char>, CMRTKernelBase*> >]' at /var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-12.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/13.0.1/bits/stl_map.h:1172:18,
    inlined from 'virtual MOS_STATUS CodechalFeiHevcStateG9Skl::FreeEncResources()' at ../media_driver/agnostic/gen9_skl/codec/hal/codechal_fei_hevc_g9_skl.cpp:6414:24:
/var/media/DATA/home-rudi/LibreELEC.tv/build.LibreELEC-Generic.x86_64-12.0-devel/toolchain/x86_64-libreelec-linux-gnu/include/c++/13.0.1/bits/stl_tree.h:2092:36: error: storing the address of local variable 'deallocator' in '*MEM[(struct _Rb_tree_node_base * &)&deallocator + 16].std::_Rb_tree_node_base::_M_parent' [-Werror=dangling-pointer=]
 2092 |           __t._M_root()->_M_parent = __t._M_end();
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
../media_driver/agnostic/gen9_skl/codec/hal/codechal_fei_hevc_g9_skl.cpp: In member function 'virtual MOS_STATUS CodechalFeiHevcStateG9Skl::FreeEncResources()':
../media_driver/agnostic/gen9_skl/codec/hal/codechal_fei_hevc_g9_skl.cpp:6413:22: note: 'deallocator' declared here
 6413 |      CmKernelMapType deallocator;
      |                      ^~~~~~~~~~~
../media_driver/agnostic/gen9_skl/codec/hal/codechal_fei_hevc_g9_skl.cpp:6413:22: note: 'deallocator.std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*> > >::_M_t.std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<ch# Linker hash-style is set to gnu via gcc default
ar> >, std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*>, std::_Select1st<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*> > >::_M_impl.std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*>, std::_Select1st<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, CMRTKernelBase*> > >::_Rb_tree_impl<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, true>::<unnamed>.std::_Rb_tree_header::_M_header.std::_Rb_tree_node_base::_M_parent' declared here
cc1plus: all warnings being treated as errors
[537/1453] Building CXX object media_driver/CMakeFiles/iHD_drv_v...CODEC.dir/agnostic/gen11/codec/hal/codechal_decode_vc1_g11.cpp.o
ninja: build stopped: subcommand failed.
```

### Debug Information

_No response_

### Do you want to contribute a patch to fix the issue?

None

### LINKED ISSUE #1641: [Bug]: build of 23.1.5 with gcc 13 fails with array-bounds
### Which component impacted?

_No response_

### Is it regression? Good in old configuration?

None

### What happened?

Build media-driver with gcc-13-20230402.
builds fine with gcc-12.2

### What's the usage scenario when you are seeing the problem?

Others

### What impacted?

Build fails using gcc-13 20230402
``` 
In member function 'void encode::BufferQueue::SetResourceType(encode::ResourceType)',
    inlined from 'std::shared_ptr<encode::BufferQueue> encode::TrackedBuffer::GetBufferQueue(encode::BufferType)' at ../media_softlet/agnostic/common/codec/hal/enc/shared/bufferMgr/encode_tracked_buffer.cpp:281:31:
../media_softlet/agnostic/common/codec/hal/enc/shared/bufferMgr/encode_tracked_buffer_queue.h:90:65: error: array subscript 0 is outside array bounds of 'std::aligned_storage<184, 8>::type [0]' [-Werror=array-bounds=]
   90 |     void SetResourceType(ResourceType resType) { m_resourceType = resType; }
      |                                                  ~~~~~~~~~~~~~~~^~~~~~~~~
In member function 'std::shared_ptr<encode::BufferQueue> encode::TrackedBuffer::GetBufferQueue(encode::BufferType)':
cc1plus: note: source object is likely at address zero
cc1plus: all warnings being treated as errors
``` 

### Debug Information

_No response_

### Do you want to contribute a patch to fix the issue?

None

### PR DESCRIPTION
**Pull request by MicroYY**
Fix #1639
Fix #1640
Fix #1641
Signed-off-by: Jay Yang <jay.yang@intel.com>

### CODE REVIEW

**Comment by heitbaum** [comment]
> Fix #1639
> Fix #1640
> Fix #1641
> Signed-off-by: Jay Yang [jay.yang@intel.com](mailto:jay.yang@intel.com)

With gcc-13.1.0-RC2-20230419

Tested-by: Rudi Heitbaum <rudi@heitbaum.com>

File: media_driver/media_softlet/agnostic/gen12_base/codec/hal/dec/shared/hucitf/decode_huc_packet_creator_g12.h [hunk 1]
```diff
@@ -48,6 +48,10 @@ class HucPacketCreatorG12 : public HucPacketCreatorBase
         MediaPipeline       *pipeline,
         MediaTask           *task,
         CodechalHwInterface *hwInterface);
+        
+    using HucPacketCreatorBase::CreateHucCopyPkt;
+    using HucPacketCreatorBase::CreateProbUpdatePkt;
+    using HucPacketCreatorBase::CreateStreamOutInterface;
```
Comment (Jexu) [inline media_driver/media_softlet/agnostic/gen12_base/codec/hal/dec/shared/hucitf/decode_huc_packet_creator_g12.h:54]: @StevenLi01 Could you help comment, why we define and use CodechalHwInterfaceNext in base class, but change it to CodechalHwInterface in child class and use legacy CodechalHwInterface everywhere in APO code?
Override class functions with different type of inputs seems not allowed by gcc13 now.

**Comment by xry111** [comment]
Any status update?

**Comment by MicroYY** [comment]
@XinfengZhang  Any comment?

**Comment by renodr** [comment]
Any further updates on this so far?

**Review by XinfengZhang — APPROVED** [review]
TBH, it is not a elegant one, but enough to pass gen13 request. maybe could revisit and have a better one in future.

**Pull request merged**